### ISSUE TITLE
docker daemon causing error with registry v2

### ISSUE THREAD

**Issue by rafabene** [docker/docker#14098]
### Description of problem:

I have a registry:2.0.1 running on a docker-machine host (provider=virtualbox with port forwarding). 
On a remote host, When I run docker pull `<remote_host>:5000/<my_image>` I got the following output:

```
latest: Pulling from <remote_host>:5000/<my_image>
8ebbde154a62: Pulling fs layer 
c12da83ee7c1: Pulling fs layer 
a16207605ac6: Pulling fs layer 
ea994adb8f5e: Pulling fs layer 
b9a671861279: Pulling fs layer 
a58de78a6a7f: Pulling fs layer 
12724ce38086: Pulling fs layer 
6bafb24bd6c1: Pulling fs layer 
4d217892bcfe: Pulling fs layer 
b741ec2037b8: Pulling fs layer 
983765c3d97c: Pulling fs layer 
96f8e8f069ed: Pulling fs layer 
Pulling repository <remote_host>:5000/<my_image>
FATA[0000] Error: image <my_image> not found.
```

Initially I though that it was an error on registry and for that reason I raised an issue at https://github.com/docker/distribution/issues/639

After some debug sessions and instructed by @stevvooe I'm reporting this bug at https://github.com/docker/docker

More information (docker dameon's log) about the issue can be found at https://gist.github.com/rafabene/64296e23cfd8ab572587 or at https://github.com/docker/distribution/issues/639
### docker version:

```
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```
### docker info:

```
Containers: 0
Images: 0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.5-boot2docker
Operating System: Boot2Docker 1.7.0 (TCL 6.3); master : 7960f90 - Thu Jun 18 18:31:45 UTC 2015
CPUs: 1
Total Memory: 996.2 MiB
Name: lab
ID: PDZY:FACX:O67C:VT5B:TULR:F6O5:BAHV:WB3T:YFVK:VAWH:GCMI:IJRP
Debug mode (server): true
File Descriptors: 10
Goroutines: 15
System Time: 2015-06-22T15:06:20.507139071Z
EventsListeners: 0
Init SHA1: 
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Labels:
 provider=virtualbox
```
### uname -a:

```
Linux Colosus 4.0.4-303.fc22.x86_64 #1 SMP Thu May 28 12:37:06 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```
### Environment details:

```
Running VirtualBox 4.3.26r98988  created by docker-machine version 0.3.0 (0a251fe)
```

**Comment by calavera** [comment]
I've removed this issue from 1.7.1 release since it looks like it was already present in 1.6.2. It can still be released in 1.7.1 if we find the root cause and we can fix it, but otherwise, this should not block the release.

**Comment by sayash** [comment]
I think I'm seeing a similar issue. It was reported over at docker/distribution issue #468 and stevvooe recommended this be posted here.

This is the message from the Docker client when pulling an image:
```
latest: Pulling from x.y.z:5000/kcontrol  
562fccfb6d84: Pull complete
cecc1f2b656b: Pull complete  
82e59b0bcc13: Pull complete
cb77c9a66b88: Download complete  
626d5a81290d: Download complete  
83e4dde6b9cf: Already exists  
b670fb0c7ecd: Already exists
29460ac93442: Already exists  
d2a0ecffe6fa: Already exists
0446758af79d: Already exists
2c93318dc1f9: Already exists                          
4e82b4cb1ec8: Already exists
Pulling repository x.y.z:5000/kcontrol  
Error: image kcontrol:latest not found
```

If I look at the logs of the registry, this is what I see:

## Registry Logs

```
172.19.2.93 - - [14/Jul/2015:21:55:41 +0000] "GET /v2/kcontrol/blobs/sha256:db797c3f390882d0ffa4384179a343eb5b340023272bc8d886e90e702cb43563 HTTP/1.1" 200 90 "" "docker/1.7.0 go/go1.4.2 git-commit/0baf609 kernel/4.0.5-boot2docker os/linux arch/amd64"
time="2015-07-14T21:55:41.016677367Z" level=info msg="filesystem.ReadStream(\"/docker/registry/v2/blobs/sha256/10/107c99f248e2939f2f73b97f75667d2a0518b660df26350fc1b6d2476c5976f0/data\", 0)" trace.duration=43.602µs trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).ReadStream" trace.id=7ba711d2-8c35-413c-9e63-04948f82b15f trace.line=79
time="2015-07-14T21:55:41.191180162Z" level=info msg="response completed" environment=development http.request.host="csig.ddns.net:5000" http.request.id=2fe28f3e-46b5-4800-b840-d80de3d7bf92 http.request.method=GET http.request.remoteaddr="172.19.2.93:46494" http.request.uri="/v2/kcontrol/blobs/sha256:107c99f248e2939f2f73b97f75667d2a0518b660df26350fc1b6d2476c5976f0" http.request.useragent="docker/1.7.0 go/go1.4.2 git-commit/0baf609 kernel/4.0.5-boot2docker os/linux arch/amd64" http.response.contenttype="application/x-gzip" http.response.duration=185.849234ms http.response.status=200 http.response.written=6018204 instance.id=872df3d1-616e-4242-be3d-890e3eb88214 service=registry version=v2.0.1

172.19.2.93 - - [14/Jul/2015:21:55:41 +0000] "GET /v2/kcontrol/blobs/sha256:107c99f248e2939f2f73b97f75667d2a0518b660df26350fc1b6d2476c5976f0 HTTP/1.1" 200 6018204 "" "docker/1.7.0 go/go1.4.2 git-commit/0baf609 kernel/4.0.5-boot2docker os/linux arch/amd64"
time="2015-07-14T21:55:43.029150689Z" level=info msg="response completed" environment=development http.request.host="csig.ddns.net:5000" http.request.id=3fcab66b-ae5b-415e-96c2-005dff17e10b http.request.method=GET http.request.remoteaddr="172.19.2.93:46499" http.request.uri="/v1/repositories/kontrol/images" http.request.useragent="docker/1.7.0 go/go1.4.2 git-commit/0baf609 kernel/4.0.5-boot2docker os/linux arch/amd64" http.response.contenttype="text/plain; charset=utf-8" http.response.duration="103.962µs" http.response.status=404 http.response.written=19 instance.id=872df3d1-616e-4242-be3d-890e3eb88214 service=registry version=v2.0.1

172.19.2.93 - - [14/Jul/2015:21:55:43 +0000] "GET /v1/repositories/kcontrol/images HTTP/1.1" 404 19 "" "docker/1.7.0 go/go1.4.2 git-commit/0baf609 kernel/4.0.5-boot2docker os/linux arch/amd64
```

## Docker Version

```
docker@admin:/mnt/sda1$ docker version
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```
## Docker Info

```
cker@admin:/mnt/sda1$ docker info
Containers: 3
Images: 30
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.5-boot2docker
Operating System: Boot2Docker 1.7.0 (TCL 6.3); master : 7960f90 - Thu Jun 18 18:31:45 UTC 2015
CPUs: 2
Total Memory: 1.956 GiB
Name: admin
ID: EI4F:BPCO:ZWDF:3MPZ:Q576:E2PK:HOFC:4IRF:Z5ZR:2EG5:26IM:MFW7
Debug mode (server): true
File Descriptors: 23
Goroutines: 37
System Time: 2015-07-14T23:57:44.522288455Z
EventsListeners: 0
Init SHA1: 
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Labels:
 provider=virtualbox
```

## uname -a

```
docker@admin:/mnt/sda1$ uname -a
Linux admin 4.0.5-boot2docker #1 SMP Tue Jun 16 01:39:56 UTC 2015 x86_64 GNU/Linux
```

The Docker VM was created with docker-machine v0.3 on VirtualBox.

**Comment by mpchlets** [comment]
I am seeing this same issue with

registry:2 (b4ad0b763f11)
Docker version 1.6.2, build 7c8fca2-dirty

I am backing the registry with S3 and have isolated down to 1 machine. 

The symptoms are the same, starts downloading, then finishes with image not found.

I am more than happy to debug as this is somewhat persistent in my setup.

**Comment by stevvooe** [comment]
This is a little tenuous, but this might be related to a fix we have in #15040.

@mpchlets @rafabene The fix should be in the next docker 1.8 release candidate. Would you mind giving it a try when this is available? If you can't wait and you're capable of building docker, you can checkout that PR branch and give the fix a try.

Reasoning: problems that kicked off #15040 were characterized by several successful v2 fetches followed by an EOF then fallback to v1. From the logs in the original complaint:

```
time="2015-06-20T00:14:58.318469285Z" level=error msg="Error from V2 registry: Get http://classroom.example.com:5000/v2/mysql/blobs/sha256:fdd04e6fcc43523cb5d1f8f09467d7b30925b7a2c06e86956fe0ba85a9dced6b: EOF" 
time="2015-06-20T00:14:58.318497762Z" level=debug msg="image does not exist on v2 registry, falling back to v1" 
time="2015-06-20T00:14:58.318508996Z" level=debug msg="pulling v1 repository with local name \"classroom.example.com:5000/mysql\"" 
```

The issue is a "double decompress" wipes out a reader, which then returns EOF on a later read.

**Comment by thaJeztah** [comment]
ping @rafabene @mpchlets could you try if you are still able to reproduce this on the current release candidate? You can find 1.8.0-rc3 here; https://github.com/docker/docker/releases/tag/v1.8.0-rc3 thanks!

**Comment by mpchlets** [comment]
I just tried with the v1.8.1 release and I am still experiencing an issue but perhaps different or just manifesting differently, it unable to write the btrfs layer - could this just now be a slow disk?

**Comment by vbichov** [comment]
getting a similar issue (see below) at AWS with Docker 1.8.1 and Distribution 2.1.1
however, it is interesting (and may or may not be helpful) that this issue does NOT occur when a pulling instance has a public IP that is exposed to the internet.

getting:
`511136ea3c5a: Pulling fs layer
b6718650e87e: Pulling fs layer
3d3c8202a574: Pulling fs layer
0114405f9ff1: Pulling fs layer
909b4f6c24a3: Pulling fs layer
ad4d06fcc614: Pulling fs layer
48099a110ee0: Pulling fs layer
f923b4c4b0f7: Pulling fs layer
b69305eac730: Pulling fs layer
155bbe8ea690: Pulling fs layer
7bc0d465ab2f: Pulling fs layer
bfb6c1a40c0d: Pulling fs layer
e6b65ddbabe2: Pulling fs layer
ea6f683b3a1f: Pulling fs layer
247b7849fd43: Pulling fs layer
a4bfb897963c: Pulling fs layer
e4b0e138b197: Pulling fs layer
706ebd5d0f05: Pulling fs layer
ffd6164bcb29: Pulling fs layer
a5831dcb25a9: Pulling fs layer
6b92501ef474: Layer already being pulled by another client. Waiting.`
and seeing 404 for v1 in the logs.

**Comment by thaJeztah** [comment]
I'm going to close this issue as the original issue was resolved in 1.8, and subsequent comments were over a year old. If you're having this issue on the current version of docker, please open a new issue instead
